### PR TITLE
Make RequireResourcesProperty work with workflows

### PR DESCRIPTION
**Pull request by davidvanlaatum**
Make RequireResourcesProperty work with workflows which allows keeping a workflow in the queue until another job has finished

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -66,6 +66,11 @@
 
 	<dependencies>
 		<dependency>
+			<groupId>org.jenkins-ci.plugins.workflow</groupId>
+			<artifactId>workflow-multibranch</artifactId>
+			<version>${workflow.version}</version>
```
Comment (amuniz) [inline pom.xml]: :-1: This dependency is not needed.

**Comment by amuniz** [comment]
Sorry, but this patch does not align with the aim of Pipeline, which is having as much as possible the job defined inside an inline Pipeline script (or Jenkinsfile in the repository).

By the way, your use case can be achieved by using a top level `lock` in your Pipeline script. It's true that the build will not wait for locking the resource in the queue but it uses a `FlyweightTask`, which does not consume an executor anyway.

So I'm :-1: on this change.

**Comment by davidvanlaatum** [comment]
I need the build to block as otherwise I end up with 20 aborted builds to one complete which looses the the info about who broke a unit test and is overall useless. The tests take around an hour and with commits being done to multiple branches just after a release it can be 2-3 hours before it gets to it. Using the inline lock also screws up runtime stats as it makes it look like it took 3 hours when 2 hours it was waiting for the lock

**Comment by amuniz** [comment]
> Using the inline lock also screws up runtime stats as it makes it look like it took 3 hours when 2 hours it was waiting for the lock

You can use `stage`s (+ stage view plugin) to properly configure your script in order to differentiate waiting stages from real work stages, so you have all timing information.

**Comment by davidvanlaatum** [comment]
I am aware of that but doesn't provide all the info such as a graph and you provided no solution to the other part?

File: pom.xml (hunk 1 above)
Comment (davidvanlaatum) [inline pom.xml]: Removed

**Comment by davidvanlaatum** [comment]
https://gist.github.com/davidvanlaatum/3bfb4bb4adc4a63653da17a5e2a4a546 is my jenkinsfile

**Comment by davidvanlaatum** [comment]
I am happy to take a different approach if it avoid the problem but I have spent a number of hours on this and this is the only thing I have found that works properly. (Using the build blocker plugin seems to hit a bug and builds get stuck and have been unable to replicate with a smaller project)

**Comment by amuniz** [comment]
> I need the build to block as otherwise I end up with 20 aborted builds to one complete which looses the the info about who broke a unit test and is overall useless

20 aborted builds to one complete? Can you elaborate a bit more please? `lock` does not abort builds.

**Comment by davidvanlaatum** [comment]
No but pipeline collapses the builds that are in duplicate stages if you don't allow concurrency which show up as aborted. So the first build is waiting on the lock second build waits for the first to exit that stage, third detects number 2 is at the same point and aborts it leaving just 3 waiting

**Comment by amuniz** [comment]
`stage concurrency: 1` don't mix well with `lock`. Actually, `concurrency` configuration will be deprecated as soon as https://github.com/jenkinsci/pipeline-milestone-step-plugin/pull/1 is merged and released.

You can see an example using `lock`, `milestone` and the new block-scope `stage` in https://github.com/jenkinsci/workflow-aggregator-plugin/pull/3

**Comment by davidvanlaatum** [comment]
I don't see how that solves my problem sounds like I will still end up with either a large backlog of builds for single commits or lots of aborted builds

**Comment by amuniz** [comment]
Sorry @davidvanlaatum, I still don't see what your use case is. If it is a legitimate use case and the only way to achieve it is by merging this PR, it will be merged, of course.

So, going back to your use case, If you don't want builds aborted, just remote the `concurrency` configuration from your `stage`s and replace it with:

```groovy
stage 'Checkout'
lock('resource1') { //add `inversePrecedence: true` if you want the newest build to be pick up from the waiting queue
  ...
}
```

BTW, if you end up using fictitious resources (like in the example above) then it sounds like you are not really in need of shared resources to lock, but just non-concurrent builds, which can be achieved by disabling the "Execute concurrent builds" check. 
In that case what you need is [JENKINS-34547](https://issues.jenkins-ci.org/browse/JENKINS-34547) if you are using a Multibranch project (or just disable it in the UI if using plain Pipeline jobs).

**Comment by davidvanlaatum** [comment]
In my case the shared resource is an apache virtualhost and database

**Comment by davidvanlaatum** [comment]
and Execute concurrent builds cannot be unchecked on multi branch pipelines to my knowledge

**Comment by amuniz** [comment]
> In my case the shared resource is an apache virtualhost and database

Great, then I guess you don't need to lock on `stage 'Checkout'`, and adding individual `lock`s steps where you need mutual exclusion would be ok, right?

> and Execute concurrent builds cannot be unchecked on multi branch pipelines to my knowledge

Yes, JENKINS-34547 is needed, as said before. Maybe your PR should go that way?

**Comment by davidvanlaatum** [comment]
Ops missed that last sentence ;) but untill that is fixed

**Comment by davidvanlaatum** [comment]
But locks would cause every commit to kick of a new build causing a big backlog

**Comment by amuniz** [comment]
> But locks would cause every commit to kick of a new build causing a big backlog

What's wrong with that? All that waiting builds will consume nothing in Jenkins as they are `FlyweightTask`s

**Comment by davidvanlaatum** [comment]
so far this week 234 so ave 58 per day.
1 hour runtime. 
168 hours per week.
234-168=66
so a build queue growing by 66 per week is not a problem in your eyes?

**Comment by davidvanlaatum** [comment]
Even if they consume no memory they will use disk space and cause the build list to be almost useless and eventually we will hit a limit not to mention all the extra power it will use working 24/7 when it doesn't need to

**Pull request closed without merging**